### PR TITLE
Fix 14455: _opt_relrotarg_valref test failure on amd64.

### LINKED ISSUE #14455: Test fails with JitStress=2: _opt_relrotarg_valref
With JitStress=2, this test fails on x64 (it passes on x86):
JIT\Methodical\explicit\rotate\_opt_relrotarg_valref

### PR DESCRIPTION
**Pull request by sandreenko**
Do not zero-initialize temps, that were created by jit and do not have GC refs.
The confusion was that `varDsc->lvIsTemp` means `short live` variable, when we wanted to ask is it IL temp or not.

Fix #14455.

### CODE REVIEW

**Comment by sandreenko** [comment]
Maybe somebody knows how to rename `lvIsTemp` (`lvShortLive`?) and `compLocalsCount`  to make things clearer.

**Comment by sandreenko** [comment]
I am trying to write a test that do not need stress modes to repro this issue.

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT jitstress2
@dotnet-bot test Windows_NT x64 corefx_jitstress2

**Comment by sandreenko** [comment]
@dotnet-bot test Ubuntu armlb Cross Debug Innerloop Build

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT x86 Checked jitstress2

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -8109,22 +8109,23 @@ void Compiler::fgMorphRecursiveFastTailCallIntoLoop(BasicBlock* block, GenTreeCa
...
         unsigned   varNum;
         LclVarDsc* varDsc;
         for (varNum = 0, varDsc = lvaTable; varNum < lvaCount; varNum++, varDsc++)
         {
-            var_types lclType = varDsc->TypeGet();
             if (!varDsc->lvIsParam)
             {
-                if (!varDsc->lvIsTemp || ((lclType == TYP_STRUCT) && (varDsc->lvStructGcCount > 0)))
+                var_types lclType            = varDsc->TypeGet();
+                bool      isILTemp           = (varNum < info.compLocalsCount);
```
Comment (AndyAyersMS) [inline src/jit/morph.cpp]: Shouldn't this variable be named `isNotILTemp` ?

**Comment by AndyAyersMS** [comment]
Adding a direct link to the bug: #14455.

Would like to see an analysis of the diffs this causes.

Let me know if you need help working up a non-stress repro.

Note not zeroing temp structs may provoke more cases of #1305. Our position is that the values of padding voids in structs are undefined there is still code out there that looks at them.

File: src/jit/morph.cpp (hunk 1 above)
Comment (sandreenko) [reply src/jit/morph.cpp]: If we call local variables from C#(or IL) as temp IL variables, then it should be `isILTemp`.
`lvaTable` starts with IL method arguments, that we skip, than IL temps, that we have to zero-initialize.

Comment (erozenfeld) [reply src/jit/morph.cpp]: This name is confusing. I would prefer isUserLocal.

**Comment by sandreenko** [comment]
So to repro this problem run the test without stress mode, but change `void Compiler::fgAdjustForAddressExposedOrWrittenThis()` to create a temp for the modifiacle this inside `growTree` method:
`if (verbose) lvaTable[info.compThisArg].lvHasILStoreOp = true;`

The problem with non-stress test case, that we can't have `lvAddrExposed` on this because it will break tail call optimization: 
```
            if (hasAddrExposedVars)
            {
                szFailReason = "Local address taken";
            }
```
and we can't do `lvHasILStoreOp` with this as well, because even in IL this address is read-only.

**Comment by sandreenko** [comment]
However, if `fgAdjustForAddressExposedOrWrittenThis` checks `lvaTable[info.compThisArg].lvHasILStoreOp` there should be a way to force it without stress modes, does anybody know how to do it?

**Comment by sandreenko** [comment]
The repro was added.

File: src/jit/morph.cpp (hunk 1 above)
Comment (sandreenko) [reply src/jit/morph.cpp]: Fixed.

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT jitstress2
@dotnet-bot test Windows_NT x86 Checked jitstress2

**Comment by sandreenko** [comment]
badAsm before:
```
00007FFB0B970CDE  add         byte ptr [rax],al  
00007FFB0B970CE0  push        rdi  
00007FFB0B970CE1  push        rsi  
00007FFB0B970CE2  push        rbx  
00007FFB0B970CE3  sub         rsp,20h  
00007FFB0B970CE7  mov         esi,edx  
00007FFB0B970CE9  mov         rdi,rcx  
00007FFB0B970CEC  test        esi,esi  
00007FFB0B970CEE  jle         00007FFB0B970D51  
00007FFB0B970CF0  mov         rcx,7FFB0B8564A0h  
00007FFB0B970CFA  call        JIT_New (07FFB6A908AD0h)  
00007FFB0B970CFF  mov         rbx,rax  
00007FFB0B970D02  mov         rcx,rbx  
00007FFB0B970D05  call        00007FFB65DCEBE0  
00007FFB0B970D0A  lea         rcx,[rdi+8]  
00007FFB0B970D0E  mov         rdx,rbx  
00007FFB0B970D11  call        JIT_WriteBarrier (07FFB6ACCFA00h)  
00007FFB0B970D16  mov         rcx,qword ptr [rdi+8]  
00007FFB0B970D1A  dec         esi  
00007FFB0B970D1C  mov         edx,esi  
00007FFB0B970D1E  cmp         dword ptr [rcx],ecx  
00007FFB0B970D20  call        00007FFB0B970898  
00007FFB0B970D25  mov         rcx,7FFB0B8564A0h  
00007FFB0B970D2F  call        JIT_New (07FFB6A908AD0h)  
00007FFB0B970D34  mov         rbx,rax  
00007FFB0B970D37  mov         rcx,rbx  
00007FFB0B970D3A  call        00007FFB65DCEBE0  
00007FFB0B970D3F  lea         rcx,[rdi+10h]  
00007FFB0B970D43  mov         rdx,rbx  
00007FFB0B970D46  call        JIT_WriteBarrier (07FFB6ACCFA00h)  
00007FFB0B970D4B  xor         edi,edi  
00007FFB0B970D4D  test        esi,esi  
00007FFB0B970D4F  jg          00007FFB0B970CF0  
00007FFB0B970D51  xor         eax,eax  
00007FFB0B970D53  mov         qword ptr [rdi+10h],rax  
00007FFB0B970D57  mov         qword ptr [rdi+8],rax  
00007FFB0B970D5B  add         rsp,20h  
00007FFB0B970D5F  pop         rbx  
00007FFB0B970D60  pop         rsi  
00007FFB0B970D61  pop         rdi  
00007FFB0B970D62  ret  
```

good asm after:
```
00007FFB0B970CDE  add         byte ptr [rax],al  
00007FFB0B970CE0  push        rdi  
00007FFB0B970CE1  push        rsi  
00007FFB0B970CE2  push        rbx  
00007FFB0B970CE3  sub         rsp,20h  
00007FFB0B970CE7  mov         esi,edx  
00007FFB0B970CE9  mov         rdi,rcx  
00007FFB0B970CEC  test        esi,esi  
00007FFB0B970CEE  jle         00007FFB0B970D5C  
00007FFB0B970CF0  mov         rcx,7FFB0B8564A0h  
00007FFB0B970CFA  call        JIT_New (07FFB6A908AD0h)  
00007FFB0B970CFF  mov         rbx,rax  
00007FFB0B970D02  mov         rcx,rbx  
00007FFB0B970D05  call        00007FFB65DCEBE0  
00007FFB0B970D0A  lea         rcx,[rdi+8]  
00007FFB0B970D0E  mov         rdx,rbx  
00007FFB0B970D11  call        JIT_WriteBarrier (07FFB6ACCFA00h)  
00007FFB0B970D16  mov         rcx,qword ptr [rdi+8]  
00007FFB0B970D1A  dec         esi  
00007FFB0B970D1C  mov         edx,esi  
00007FFB0B970D1E  cmp         dword ptr [rcx],ecx  
00007FFB0B970D20  call        00007FFB0B970898  
00007FFB0B970D25  mov         rcx,7FFB0B8564A0h  
00007FFB0B970D2F  call        JIT_New (07FFB6A908AD0h)  
00007FFB0B970D34  mov         rbx,rax  
00007FFB0B970D37  mov         rcx,rbx  
00007FFB0B970D3A  call        00007FFB65DCEBE0  
00007FFB0B970D3F  lea         rcx,[rdi+10h]  
00007FFB0B970D43  mov         rdx,rbx  
00007FFB0B970D46  call        JIT_WriteBarrier (07FFB6ACCFA00h)  
00007FFB0B970D4B  mov         rcx,qword ptr [rdi+10h]  
00007FFB0B970D4F  mov         rdi,rcx  
00007FFB0B970D52  mov         rcx,rdi  
00007FFB0B970D55  mov         rdi,rcx  
00007FFB0B970D58  test        esi,esi  
00007FFB0B970D5A  jg          00007FFB0B970CF0  
00007FFB0B970D5C  xor         eax,eax  
00007FFB0B970D5E  mov         qword ptr [rdi+10h],rax  
00007FFB0B970D62  mov         qword ptr [rdi+8],rax  
00007FFB0B970D66  add         rsp,20h  
00007FFB0B970D6A  pop         rbx  
00007FFB0B970D6B  pop         rsi  
00007FFB0B970D6C  pop         rdi  
00007FFB0B970D6D  ret  
```

so the diff:
```
--- C:\Users\seandree\Documents\work\november\badAsm.txt
+++ C:\Users\seandree\Documents\work\november\goodAsm.txt
@@ -6,7 +6,7 @@
 00007FFB0B970CE7  mov         esi,edx  
 00007FFB0B970CE9  mov         rdi,rcx  
 00007FFB0B970CEC  test        esi,esi  
-00007FFB0B970CEE  jle         00007FFB0B970D51  
+00007FFB0B970CEE  jle         00007FFB0B970D5C  
 00007FFB0B970CF0  mov         rcx,7FFB0B8564A0h  
 00007FFB0B970CFA  call        JIT_New (07FFB6A908AD0h)  
 00007FFB0B970CFF  mov         rbx,rax  
@@ -28,14 +28,17 @@
 00007FFB0B970D3F  lea         rcx,[rdi+10h]  
 00007FFB0B970D43  mov         rdx,rbx  
 00007FFB0B970D46  call        JIT_WriteBarrier (07FFB6ACCFA00h)  
-00007FFB0B970D4B  xor         edi,edi  
-00007FFB0B970D4D  test        esi,esi  
-00007FFB0B970D4F  jg          00007FFB0B970CF0  
-00007FFB0B970D51  xor         eax,eax  
-00007FFB0B970D53  mov         qword ptr [rdi+10h],rax  
-00007FFB0B970D57  mov         qword ptr [rdi+8],rax  
-00007FFB0B970D5B  add         rsp,20h  
-00007FFB0B970D5F  pop         rbx  
-00007FFB0B970D60  pop         rsi  
-00007FFB0B970D61  pop         rdi  
-00007FFB0B970D62  ret  +00007FFB0B970D4B  mov         rcx,qword ptr [rdi+10h]  
+00007FFB0B970D4F  mov         rdi,rcx  
+00007FFB0B970D52  mov         rcx,rdi  
+00007FFB0B970D55  mov         rdi,rcx  
+00007FFB0B970D58  test        esi,esi  
+00007FFB0B970D5A  jg          00007FFB0B970CF0  
+00007FFB0B970D5C  xor         eax,eax  
+00007FFB0B970D5E  mov         qword ptr [rdi+10h],rax  
+00007FFB0B970D62  mov         qword ptr [rdi+8],rax  
+00007FFB0B970D66  add         rsp,20h  
+00007FFB0B970D6A  pop         rbx  
+00007FFB0B970D6B  pop         rsi  
+00007FFB0B970D6C  pop         rdi  
+00007FFB0B970D6D  ret 
``` 
bad `xor         edi,edi` was not inserted. New moves in the good version are correct, they were deleted by dce in the bad version, it is why the good version is longer.

**Comment by sandreenko** [comment]
SPMI showed small improvements, several movs were deleted, all changes were as expected.
<filename>  baseline           diff    improvement      %improvement    #funcs
TOTAL             10617       10467                    150                       1.41           20

**Comment by sandreenko** [comment]
ci failures are not related with this change.

**Comment by sandreenko** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test 
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test
@dotnet-bot test Ubuntu16.04 armlb Cross Debug Innerloop Build

**Pull request merged**